mp4parse/src/lib.rs

```diff
@@ -388,6 +392,529 @@ pub enum SampleEntry {
+                .and_then(|res| {
+                    self.remain_chunk_count = res;
+                    self.chunks.next()
+                })
+        });
+
+        has_chunk.map(|id| (id, self.sample_count))
+    }
+}
+
+impl<'a> SampleToChunkIterator<'a> {
+    #[allow(clippy::reversed_empty_ranges)]
+    fn locate(&mut self) -> std::ops::Range<u32> {
+        loop {
+            return match (self.stsc_peek_iter.next(), self.stsc_peek_iter.peek()) {
+                (Some(next), Some(peek)) if next.first_chunk == peek.first_chunk => {
+                    // Invalid entry, skip it and will continue searching at
+                    // next loop iteration.
+                    continue;
+                }
+                (Some(next), Some(peek)) if next.first_chunk > 0 && peek.first_chunk > 0 => {
+                    self.sample_count = next.samples_per_chunk;
+                    (next.first_chunk - 1)..(peek.first_chunk - 1)
+                }
+                (Some(next), None) if next.first_chunk > 0 => {
+                    self.sample_count = next.samples_per_chunk;
+                    // Total chunk number in 'stsc' could be different to 'stco',
+                    // there could be more chunks at the last 'stsc' record.
+                    match next.first_chunk.checked_add(self.remain_chunk_count) {
+                        Some(r) => (next.first_chunk - 1)..r - 1,
+                        _ => 0..0,
+                    }
+                }
+                _ => 0..0,
+            };
+        }
+    }
+}
+
+/// Calculate numerator * scale / denominator, if possible.
+///
+/// Applying the associativity of integer arithmetic, we divide first
+/// and add the remainder after multiplying each term separately
+/// to preserve precision while leaving more headroom. That is,
+/// (n * s) / d is split into floor(n / d) * s + (n % d) * s / d.
+///
+/// Return None on overflow or if the denominator is zero.
+pub fn rational_scale<T, S>(numerator: T, denominator: T, scale2: S) -> Option<T>
+where
+    T: PrimInt + Zero,
+    S: PrimInt,
+{
+    if denominator.is_zero() {
+        return None;
+    }
+
+    let integer = numerator / denominator;
+    let remainder = numerator % denominator;
+    num_traits::cast(scale2).and_then(|s| match integer.checked_mul(&s) {
+        Some(integer) => remainder
+            .checked_mul(&s)
+            .and_then(|remainder| (remainder / denominator).checked_add(&integer)),
+        None => None,
+    })
+}
+
+pub fn media_time_to_us(time: MediaScaledTime, scale: MediaTimeScale) -> Option<u64> {
+    let microseconds_per_second = 1_000_000;
+    rational_scale::<u64, u64>(time.0, scale.0, microseconds_per_second)
+}
+
+pub fn track_time_to_us<T>(time: TrackScaledTime<T>, scale: TrackTimeScale<T>) -> Option<T>
+where
+    T: PrimInt + Zero,
+{
+    assert_eq!(time.1, scale.1);
+    let microseconds_per_second = 1_000_000;
+    rational_scale::<T, u64>(time.0, scale.0, microseconds_per_second)
+}
```

Comment:
This is a move of https://github.com/mozilla/mp4parse-rust/blob/1ab30201ad06389d0d0ebe3036418d624e7219f4/mp4parse_capi/src/lib.rs#L707-L746 with the addition of making these three functions public.

Generally, the same points as with `create_sample_table` apply, so I think we could use the same approaches, but it's less clear to me why a consumer of this crate would want to expose *these* functions. Can you talk more about your use case?

Also, I think if we're going to make time conversion functions part of a public interface, it would be preferable to (rather than using `u64` or `T: PrimInt + Zero`) either:
- Make use of a newtype like `pub struct Microseconds<T>(T);` which makes unit-mixing mistakes easier to avoid
- Return a stdlib [Duration](https://doc.rust-lang.org/stable/std/time/struct.Duration.html) value

Reply:
Ok, you're right, `rational_scale` doesn't need to be public. The other two are public because they are used by `mp4parse_capi` crate. My usecase doesn't involve them directly (yet), but I do see them being useful for the consumer of this crate.

As for the return type, I think stdlib `Duration` is a better choice and I can go with that, but then should we also rename the functions? `track_time_to_duration`? or just leave as is?

Reply:
> As for the return type, I think stdlib `Duration` is a better choice and I can go with that, but then should we also rename the functions? `track_time_to_duration`? or just leave as is?

If it's not necessary to expose `media_time_to_us` and `track_time_to_us` for your use case, I'll leave it up to you to choose:
1. Keep them private
2. Change the return on the public (still behind `unstable-api` feature) functions to `Duration` (renaming them `track_time_to_duration` and `media_time_to_duration`), and leave `mp4parse_capi::{media_time_to_us, track_time_to_us}` as private functions, implemented in terms of the new `Duration`-returning ones.

Reply:
Ok I went with newtype `Microseconds<T>` because stdlib `Duration` uses `u128` to represent micros and I didn't want to add that overhead